### ISSUE TITLE
Caches should be able to use alternative key builders

### ISSUE THREAD

**Issue by mheppner** [argaen/aiocache#308]
Is there an easy way to apply a key mangler? Basically, something that can be ran after the key builder, and a way to configure it on the `default` cache. The use case is for encoding long keys, since memcached only supports a key of 250 bytes. I'd like to be able to use `hashlib.sha1(<key builder output>).hexdigest()` as the keys.

Possibly related, but I think the wrong exception was thrown when I had a key length of 274:
```python
Traceback (most recent call last):
  File "/Users/mark/Documents/app/env/lib/python3.6/site-packages/aiocache/decorators.py", line 104, in set_in_cache
    await self.cache.set(key, value, ttl=self.ttl)
  File "/Users/mark/Documents/app/env/lib/python3.6/site-packages/aiocache/base.py", line 58, in _enabled
    return await func(*args, **kwargs)
  File "/Users/mark/Documents/app/env/lib/python3.6/site-packages/aiocache/base.py", line 43, in _timeout
    return await asyncio.wait_for(func(self, *args, **kwargs), timeout)
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/tasks.py", line 352, in wait_for
    return fut.result()
  File "/Users/mark/Documents/app/env/lib/python3.6/site-packages/aiocache/base.py", line 71, in _plugins
    ret = await func(self, *args, **kwargs)
  File "/Users/mark/Documents/app/env/lib/python3.6/site-packages/aiocache/base.py", line 248, in set
    res = await self._set(ns_key, dumps(value), ttl=ttl, _cas_token=_cas_token, _conn=_conn)
  File "/Users/mark/Documents/app/env/lib/python3.6/site-packages/aiocache/backends/memcached.py", line 48, in _set
    raise TypeError("memcached doesn't support float ttl")
TypeError: memcached doesn't support float ttl
```

The `ttl` for the `@cached()` function I was calling is definitely set to an integer, so maybe it picked up on an exception about the key being too large?

**Comment by argaen** [comment]
Hmmm yeah the exception is not being propagated correctly for the memcached connector. This is easy to solve.

Regarding the `key_mangler`, right now there are two ways of setting it (which is not exactly what you're asking):

1. In the `cached` decorator you have a `key_builder` param you can pass a callable: http://aiocache.readthedocs.io/en/latest/decorators.html#cached. Note that whatever you return in the callable, the cache will prefix that key with the namespace if you are using one.

2. Create a new cache class like:

```python
class MyMemcachedCache(MemcachedCache):
 
    def _build_key(self, key, namespace=None):
        # Do whatever you like here
```
Then you can set your class as the default one in your config.

I know none of the two proposed solutions is ideal and its something I wanted to work on in future versions but for now it should do the work for you.

**Comment by mheppner** [comment]
The second option works great for me, thanks! Sorry for lumping a feature request in with a bug. And fantastic work with this library! :)

**Comment by argaen** [comment]
Just published v0.7.2 with both the feature and the bug fix.

Thanks for reporting :)
